maintain at most single overlay for each view in hybrid composition

2ZeroSix opened the pull request:
Fixed rendering of single widget in multiple overlays on top of each hybrid composition PlatformView

I wasn't able to come up with any meaningful test for this scenario (besides snapshot test on android device), would appreciate any help here.

reproduction examples:

simple case in multiple_overlays branch (from original [issue](https://github.com/flutter/flutter/issues/79371)): https://github.com/2ZeroSix/flutter_android_surface_activity_destroy_bug_example/tree/multiple_overlays
<details>
<summary> screenshots </summary>

||1 view|2 views|
|-|-|-|
|before|<img src="https://user-images.githubusercontent.com/14371067/113096927-4fc74400-9220-11eb-9a85-cd54fe9474d9.jpg" width=300>|<img src="https://user-images.githubusercontent.com/14371067/113096940-55248e80-9220-11eb-8b15-480faf581664.jpg" width=300>|
|after|<img src="https://user-images.githubusercontent.com/14371067/113096947-56ee5200-9220-11eb-8575-786a21c789fc.jpg" width=300>|<img src="https://user-images.githubusercontent.com/14371067/113096952-58b81580-9220-11eb-86cc-5d84a959acb8.jpg" width=300>|

</details>


complex case in multiple_scrollable_overlays branch: https://github.com/2ZeroSix/flutter_android_surface_activity_destroy_bug_example/tree/multiple_scrollable_overlays

<details>
<summary>videos</summary>

## before
https://user-images.githubusercontent.com/14371067/113097470-22c76100-9221-11eb-86c2-bd1d5b11b5cb.mp4 
## after
https://user-images.githubusercontent.com/14371067/113097496-2ce95f80-9221-11eb-8a76-1cb0de4738af.mp4


</details>

*List which issues are fixed by this PR. You must list at least one issue.*
fixes https://github.com/flutter/flutter/issues/79371

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

blasten commented on shell/platform/android/external_view_embedder/external_view_embedder.cc:
```diff
@@ -137,7 +137,14 @@ void AndroidExternalViewEmbedder::SubmitFrame(
+        if (overlay_layers.at(view_id).empty()) {
+          overlay_layers.at(view_id).push_back(intersection_rect);
+        } else if (intersection_rect.contains(
+                       overlay_layers.at(view_id).back())) {
```
I was expecting the overlay layers to be associated with two different view_ids since there are two platform views, and the fab button intersects both.

Do you mind verifying why a single `view_id` contain the two overlay surfaces?

blasten commented:
For the testing, we use the scenario app. https://github.com/flutter/engine/blob/master/testing/scenario_app/README.md#running-for-android.  

The tests are in https://github.com/flutter/engine/blob/master/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/PlatformViewUiTests.java.

Thanks for the contribution, and let me know if you encounter any issue.

blasten commented:
For  testing, we use the scenario app. https://github.com/flutter/engine/blob/master/testing/scenario_app/README.md#running-for-android.  

The tests are in https://github.com/flutter/engine/blob/master/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenariosui/PlatformViewUiTests.java.

Thanks for the contribution, and let me know if you encounter any issue.

2ZeroSix commented:
@blasten I updated bug reproduction example https://github.com/2ZeroSix/flutter_android_surface_activity_destroy_bug_example/tree/multiple_scrollable_overlays

I found strange behavior (like a race condition) when view should jump between overlay and background layers

To fix this I just added empty fullscreen platform view.
And now I don't understand what's going on at all 🙃
<details>
<summary>video</summary>

background android view is disabled: flutter views in listview are might get clipped in region of transition between background and overlay layers
background android view is enabled: everything works as expected in this pr

master: 

https://user-images.githubusercontent.com/14371067/113258431-211b9d00-92f6-11eb-95c8-a100af539b57.mp4

this pr:

https://user-images.githubusercontent.com/14371067/113259357-3ba24600-92f7-11eb-83b9-cf24dbb78f78.mp4


</details>

chinmaygarde commented:
Do we understand the issue and/or are we making progress?

blasten commented:
@chinmaygarde Hard to say without debugging it. I will take a look shortly.

chinmaygarde commented:
@blasten Is this still on your plate? Is there someone else we can find to review this patch?

blasten commented:
@2ZeroSix I took a closer look, and this seems reasonable. Can you add some unit tests for each of the code branches?

chinmaygarde commented:
Any updates on the unit-tests?

blasten commented:
Thanks for the contribution. I will take this PR, and add the tests.
